builtins/web/blob.cpp

```diff
@@ -0,0 +1,703 @@
+    JS_STRING_SYM_PS(toStringTag, "Blob", JSPROP_READONLY),
+    JS_PS_END,
+};
+
+class StreamTask final : public api::AsyncTask {
+    Heap<JSObject *> source_;
+
+    static constexpr size_t CHUNK_SIZE = 8192;
+
+public:
+  explicit StreamTask(const HandleObject source, api::Engine *engine)  : source_(source) {
+    engine->incr_event_loop_interest();
+    handle_ = IMMEDIATE_TASK_HANDLE;
+  }
+
+  [[nodiscard]] bool run(api::Engine *engine) override {
+    JSContext *cx = engine->cx();
+    RootedObject owner(cx, streams::NativeStreamSource::owner(source_));
+    RootedObject controller(cx, streams::NativeStreamSource::controller(source_));
+    RootedValue ret(cx);
+
+    auto readers = Blob::readers(owner);
+    auto rdr = readers->lookup(source_);
+
+    if (!rdr) {
+      return false;
+    }
+
+    auto chunk = rdr->value().read(CHUNK_SIZE);
+    auto chunk_size = chunk.size();
+
+    if (chunk.empty()) {
+      if (!JS::Call(cx, controller, "close", HandleValueArray::empty(), &ret)) {
+        return false;
+      }
+
+      readers->remove(source_);
+      return cancel(engine);
+    }
+
+    RootedObject array_buffer(cx, new_array_buffer_from_span(cx, chunk));
+    if (!array_buffer) {
+      return false;
+    }
+
+    RootedObject bytes_buffer(cx, JS_NewUint8ArrayWithBuffer(cx, array_buffer, 0, chunk_size));
+    if (!bytes_buffer) {
+      return false;
+    }
+
+    RootedValueArray<1> enqueue_args(cx);
+    enqueue_args[0].setObject(*bytes_buffer);
+    if (!JS::Call(cx, controller, "enqueue", enqueue_args, &ret)) {
+      return false;
+    }
+
+    return cancel(engine);
+  }
+
+  [[nodiscard]] bool cancel(api::Engine *engine) override {
+    handle_ = INVALID_POLLABLE_HANDLE;
+    engine->decr_event_loop_interest();
+    return true;
+  }
+
+  void trace(JSTracer *trc) override { TraceEdge(trc, &source_, "source for future"); }
+};
+
+JSObject *Blob::data_to_owned_array_buffer(JSContext *cx, HandleObject self) {
+    size_t total_size = blob_size(self);
+    size_t bytes_read = 0;
+
+    return Blob::data_to_owned_array_buffer(cx, self, 0, total_size, &bytes_read);
+}
+
+JSObject* Blob::data_to_owned_array_buffer(JSContext* cx, HandleObject self, size_t offset, size_t size, size_t* bytes_read) {
+  auto blob = Blob::blob(self);
+  auto blob_size = blob->size();
+  *bytes_read = 0;
+
+  MOZ_ASSERT(offset <= blob_size);
+
+  size_t available_bytes = blob_size - offset;
+  size_t read_size = std::min(size, available_bytes);
+
+  auto span = std::span<uint8_t>(blob->data() + offset, read_size);
+  auto array_buffer = new_array_buffer_from_span(cx, span);
+  if (!array_buffer) {
+    return nullptr;
+  }
+
+  *bytes_read = read_size;
+  return array_buffer;
+}
+
+bool Blob::arrayBuffer(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+
+  JS::RootedObject promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
```

Comment:
If we weren't able to create a `Promise` object here, we'll pretty certainly also not be able to create on in `ReturnPromiseRejectedWithPendingError`. Given that, I think it makes sense to just return `false` here.
```suggestion
  if (!promise) {
    return false;
  }
```